### ISSUE TITLE
spr function, flip and rotate parameter unexpected behavior

### ISSUE THREAD

**Issue by Kyuchumimo** [nesbox/TIC-80#1696]
see:
```lua
-- title:  game title
-- author: game developer
-- desc:   short description
-- script: lua

t=0

function TIC()
	cls(13)
	spr(1+t%60//30*2,25,14,14,2,0,0,2,2) --No rotation
	spr(1+t%60//30*2,75,14,14,2,0,1,2,2) --90° rotation
	spr(1+t%60//30*2,125,14,14,2,0,2,2,2) --180° rotation
	spr(1+t%60//30*2,175,14,14,2,0,3,2,2) --270° rotation
	spr(1+t%60//30*2,25,54,14,2,0,4,2,2) --Unexpected rotation
	spr(1+t%60//30*2,75,54,14,2,0,5,2,2) --Unexpected rotation
	spr(1+t%60//30*2,125,54,14,2,0,6,2,2) --Unexpected rotation
	spr(1+t%60//30*2,175,54,14,2,0,7,2,2) --Unexpected rotation
	spr(1+t%60//30*2,25,94,14,2,0,8,2,2) --Unexpected rotation
	spr(1+t%60//30*2,75,94,14,2,0,9,2,2) --Unexpected rotation
	spr(1+t%60//30*2,125,94,14,2,0,10,2,2) --Unexpected rotation
	spr(1+t%60//30*2,175,94,14,2,0,11,2,2) --Unexpected rotation
	print("rotate=",97,1)
	print("0",36,8)
	print("1",89,8)
	print("2",141,8)
	print("3",188,8)
	print("4",36,48)
	print("5",89,48)
	print("6",141,48)
	print("7",188,48)
	print("8",36,88)
	print("9",89,88)
	print("10",136,88)
	print("11",186,88)
        print("...",212,106)
	print("and so on...",180,129)
	t=t+1
end
```
https://user-images.githubusercontent.com/74131798/142073639-83bec41c-26f2-4d8c-8c03-72cfe82a286a.mp4

**Comment by joshgoebel** [comment]
Please add some comments to source and explain what you find unexpected and what you expected to happen instead just so we are all clear.

**Comment by Kyuchumimo** [comment]
I was expecting something like this:

    0 = No rotation
    1 = 90° rotation
    2 = 180° rotation
    3 = 270° rotation
    4 = 360° rotation (same as 0)
    5 = 450° rotation (same as 1)
    6 = 540° rotation (same as 2)
    7 = 630° rotation (same as 3)
    8 = 720° rotation (same as 0)
    .... you get it, right?

Or add in the source code a modulo operator (%) that wraps the values from 0 to 3:
`rotate%4`

A similar error occurs also with the flip parameter

**Comment by potato25** [comment]
> see:
> 
> ```lua
> -- title:  game title
> -- author: game developer
> -- desc:   short description
> -- script: lua
> 
> t=0
> 
> function TIC()
> 	cls(13)
> 	spr(1+t%60//30*2,25,14,14,2,0,0,2,2) --No rotation
> 	spr(1+t%60//30*2,75,14,14,2,0,1,2,2) --90° rotation
> 	spr(1+t%60//30*2,125,14,14,2,0,2,2,2) --180° rotation
> 	spr(1+t%60//30*2,175,14,14,2,0,3,2,2) --270° rotation
> 	spr(1+t%60//30*2,25,54,14,2,0,4,2,2) --Unexpected rotation
> 	spr(1+t%60//30*2,75,54,14,2,0,5,2,2) --Unexpected rotation
> 	spr(1+t%60//30*2,125,54,14,2,0,6,2,2) --Unexpected rotation
> 	spr(1+t%60//30*2,175,54,14,2,0,7,2,2) --Unexpected rotation
> 	spr(1+t%60//30*2,25,94,14,2,0,8,2,2) --Unexpected rotation
> 	spr(1+t%60//30*2,75,94,14,2,0,9,2,2) --Unexpected rotation
> 	spr(1+t%60//30*2,125,94,14,2,0,10,2,2) --Unexpected rotation
> 	spr(1+t%60//30*2,175,94,14,2,0,11,2,2) --Unexpected rotation
> 	print("rotate=",97,1)
> 	print("0",36,8)
> 	print("1",89,8)
> 	print("2",141,8)
> 	print("3",188,8)
> 	print("4",36,48)
> 	print("5",89,48)
> 	print("6",141,48)
> 	print("7",188,48)
> 	print("8",36,88)
> 	print("9",89,88)
> 	print("10",136,88)
> 	print("11",186,88)
>         print("...",212,106)
> 	print("and so on...",180,129)
> 	t=t+1
> end
> ```
> 
>  screen-2021-11-16_15.05.10.mp4

I find that behavior to be very cool and interesting , i will experiment with it later :)

**Comment by nesbox** [comment]
Looks interesting, but I think more predictable behavior is better.
Fixed here 9c7267e, thanks.

**Comment by joshgoebel** [comment]
I'd suggest this be reviewed one last time (and hence reopening). _The docs are clear that only 4 values are supported_ here (0-3)... 4 or 20 or 2000 are simply not valid arguments.  This is essentially an enum that represents a given rotation (of a few discrete angles).  It's in no some type of _direct_ degree equivalence - **there is no reason it should wrap at all**.  If someone wants that behavior, I think they should write it into their code.

I'd suggest that a value other than 0-4 should either throw an error (which we don't tend to do for our API) or should just default back to the default, of zero. 

IE the correct patch would be:

```c
if (rotate>4) rotate = 0;
if (flip>4) flip=0;
```

Thoughts?

**Comment by joshgoebel** [comment]
@nesbox Sorry you go to this before I could even comment on it. :)

**Comment by Kyuchumimo** [comment]
there is reason it should wrap at all. This is how the binary system works in computing.

rotate=0 -> 00
rotate=1 -> 00 + 1x1 = 01
rotate=2 -> 00 + 1x2 = 10
rotate=3 -> 00 + 1x3 = 11
rotate=4 -> 00 + 1x4 = 00
rotate=5 -> 00 + 1x5 = 01
...

**Comment by joshgoebel** [comment]
> This is how the binary system works in computing.

Irrelevant in this context.

> TIC-80 is based on memory mapping

These are arguments to a function, this has nothing to do with memory mapping.

**Comment by joshgoebel** [comment]
Closed via 9c7267ee7528a75a3ceca67674618f0fe6384728.
